File: Prácticas/deepspace/src/deepspace/BetaPowerEfficientSpaceStation.java
package deepspace;

/**
 * Represents a BetaPowerEfficientSpaceStation
 * It is the same as a PowerEfficientSpaceStation but with a chance of shooting 
 * with an extra factor
 * 
 * @author Miguel Ángel Fernández Gutiérrez
 * @author Sergio Quijano Rey
 */
public class BetaPowerEfficientSpaceStation extends PowerEfficientSpaceStation {
    /**
     * Extra efficiency for beta efficient.
     */
    private final float EXTRAEFFICIENCY = 1.2f;
    
    /**
     * Random utilities.
     */
    private Dice dice;
    
    // -------------------------------------------------------------------------
    // Constructors
    // -------------------------------------------------------------------------
    
    /**
     * Class initializer.
     * @param station basic space station to convert to beta efficient
     */
    BetaPowerEfficientSpaceStation(SpaceStation station) {
        super(station);
        dice = new Dice();
    }
    
    // -------------------------------------------------------------------------
    // Setters
    // -------------------------------------------------------------------------
    
    /**
     * Makes a shot.
     * @return the shot power
     */
    @Override
    public float fire() {
        if ( dice.extraEfficiency() )
            return super.fire()*EXTRAEFFICIENCY;
        else
            return super.fire();            
    }
    
    // -------------------------------------------------------------------------
    // String representation, UI version
    // -------------------------------------------------------------------------
    
    /**
     * String representation of the object.
     * @return string representation
     */
    @Override
    public String toString() {
        String message = "(Beta)" + super.toString();
        return message;
    }
    
    /**
     * To UI.
     */
    @Override
    public BetaPowerEfficientSpaceStationToUI getUIversion() {
        return new BetaPowerEfficientSpaceStationToUI(this);
    }
}

File: Prácticas/Deepspace/lib/BetaPowerEfficientSpaceStation.rb
#encoding:utf-8

require_relative 'Dice'
require_relative 'PowerEfficientSpaceStation'
require_relative 'BetaPowerEfficientSpaceStationToUI'

module Deepspace

# Beta power efficient space station: gives more shot power than an efficient space station
#
# @author Miguel Ángel Fernández Gutiérrez, Sergio Quijano Rey
class BetaPowerEfficientSpaceStation < PowerEfficientSpaceStation

    # @!attribute [Float] extra efficiency for beta efficient
    @@EXTRAEFFICIENCY = 1.20

    # Class initializer
    # @param station [SpaceStation] basic space station to convert to beta efficient
    # --Overriden
    def initialize(station)
        super

        # @!attribute [Dice] random utilities
        @dice = Dice.new
    end

    # Makes a shot
    # @return [Float] the shot power
    # --Overriden
    def fire
        if @dice.extraEfficiency
            return @@EXTRAEFFICIENCY*super
        else
            return super
        end
    end

    # String representation
    # @return [String] string representation
    def to_s
        message = "(Beta) " + super
        return message
    end

    # To UI
    def getUIversion
        return BetaPowerEfficientSpaceStationToUI.new(self)
    end

end # class BetaPowerEfficientSpaceStation

end # module Deepspace